Benchmem

tgulacsi opened the pull request:
See https://github.com/shamaazi/antilog/commit/80df4e49ecc3b2921a57944637b3bb4f25c549c1 for the benchmarks.

Although Go is fast, logging should not have a big performance/memory penalty - or at least it should consume as less resources as possible.

tgulacsi committed:
Add more benchmarks from rs/zerolog

tgulacsi committed:
Use a map for uniqueness and reverse-append in PrependUnique

tgulacsi committed:
Write: Encode directly into Buffer for less allocations

name                             old time/op    new time/op    delta
LogEmpty-4                         1.12µs ±10%    0.38µs ± 3%  -65.92%
(p=0.002 n=6+6)
Info-4                             1.23µs ± 4%    0.52µs ±10%  -57.43%
(p=0.002 n=6+6)
ContextFields-4                    1.97µs ±24%    0.84µs ± 5%  -57.32%
(p=0.002 n=6+6)
ContextAppend-4                    1.50µs ± 2%    0.92µs ± 4%  -38.33%
(p=0.004 n=5+6)
LogFields-4                        3.51µs ± 8%    2.52µs ± 3%  -28.21%
(p=0.002 n=6+6)
LogWithNoFields-4                  2.18µs ± 4%    0.87µs ± 7%  -60.03%
(p=0.002 n=6+6)
LogWithComplexFields-4             11.7µs ± 3%    10.8µs ± 5%   -7.41%
(p=0.004 n=6+5)
LogWithComplexFieldsInContext-4    4.23µs ±12%    2.78µs ±40%  -34.20%
(p=0.004 n=6+6)

name                             old alloc/op   new alloc/op   delta
LogEmpty-4                           536B ± 0%      144B ± 0%  -73.13%
(p=0.002 n=6+6)
Info-4                               856B ± 0%      352B ± 0%  -58.88%
(p=0.002 n=6+6)
ContextFields-4                    1.91kB ± 0%    0.90kB ± 0%  -53.14%
(p=0.002 n=6+6)
ContextAppend-4                      768B ± 0%      496B ± 0%  -35.42%
(p=0.002 n=6+6)
LogFields-4                        2.53kB ± 0%    1.52kB ± 0%  -39.90%
(p=0.002 n=6+6)
LogWithNoFields-4                    813B ± 0%      343B ± 2%  -57.78%
(p=0.004 n=6+5)
LogWithComplexFields-4             4.98kB ± 0%    4.84kB ± 2%   -2.86%
(p=0.002 n=6+6)
LogWithComplexFieldsInContext-4    3.74kB ± 1%    2.18kB ± 4%  -41.77%
(p=0.016 n=4+5)

name                             old allocs/op  new allocs/op  delta
LogEmpty-4                           18.0 ± 0%       3.0 ± 0%  -83.33%
(p=0.002 n=6+6)
Info-4                               20.0 ± 0%       5.0 ± 0%  -75.00%
(p=0.002 n=6+6)
ContextFields-4                      29.0 ± 0%       7.0 ± 0%  -75.86%
(p=0.002 n=6+6)
ContextAppend-4                      27.0 ± 0%      12.0 ± 0%  -55.56%
(p=0.002 n=6+6)
LogFields-4                          52.0 ± 0%      30.0 ± 0%  -42.31%
(p=0.002 n=6+6)
LogWithNoFields-4                    19.0 ± 0%       4.0 ± 0%  -78.95%
(p=0.002 n=6+6)
LogWithComplexFields-4               73.0 ± 0%      59.0 ± 0%  -19.18%
(p=0.002 n=6+6)
LogWithComplexFieldsInContext-4      30.0 ± 0%      14.0 ± 0%  -53.33%
(p=0.002 n=6+6)

tgulacsi committed:
Make staticcheck happy

tgulacsi committed:
Benchmark map/walk lookup, switch back to walk

The for-cycle search seems to bea a/2 the time with guaranteed no alloc.
Tested only for small (len<=16) slices.

$ go test -bench=Lookup -benchmem -benchtime 100ms
goos: linux
goarch: amd64
pkg: github.com/shamaazi/antilog
BenchmarkLookupMap/1-4            253792               439 ns/op
0 B/op          0 allocs/op
BenchmarkLookupMap/2-4            216787               589 ns/op
0 B/op          0 allocs/op
BenchmarkLookupMap/3-4            175575               735 ns/op
0 B/op          0 allocs/op
BenchmarkLookupMap/4-4            112951               941 ns/op
0 B/op          0 allocs/op
BenchmarkLookupMap/5-4            120356              1069 ns/op
0 B/op          0 allocs/op
BenchmarkLookupMap/6-4            103062              1208 ns/op
0 B/op          0 allocs/op
BenchmarkLookupMap/7-4             86959              1448 ns/op
0 B/op          0 allocs/op
BenchmarkLookupMap/8-4             86560              1445 ns/op
0 B/op          0 allocs/op
BenchmarkLookupMap/9-4             68139              1554 ns/op
288 B/op          1 allocs/op
BenchmarkLookupMap/10-4            84186              1665 ns/op
291 B/op          1 allocs/op
BenchmarkLookupMap/11-4            64197              1631 ns/op
297 B/op          1 allocs/op
BenchmarkLookupMap/12-4            70647              1724 ns/op
309 B/op          1 allocs/op
BenchmarkLookupMap/13-4            64742              1790 ns/op
326 B/op          1 allocs/op
BenchmarkLookupMap/14-4            67585              1899 ns/op
577 B/op          1 allocs/op
BenchmarkLookupMap/15-4            64528              1885 ns/op
578 B/op          1 allocs/op
BenchmarkLookupWalk/1-4           462890               251 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/2-4           305868               406 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/3-4           248193               508 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/4-4           198852               613 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/5-4           184926               701 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/6-4           154459               844 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/7-4           123939               849 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/8-4           136122              1051 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/9-4           114349              1019 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/10-4                   91245              1170 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/11-4                  108969              1154 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/12-4                   82257              1224 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/13-4                  105072              1259 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/14-4                   91714              1345 ns/op
0 B/op          0 allocs/op
BenchmarkLookupWalk/15-4                   82077              1310 ns/op
0 B/op          0 allocs/op
PASS
ok      github.com/shamaazi/antilog     94.660s

tgulacsi committed:
Move lookup benchmarks under "lookup" tag

tgulacsi committed:
PrependUnique: remove reverse-append-reverse

Use a big enough slice with a decreasing insertion point.
Simpler and a little bit faster:

name                             old time/op    new time/op    delta
LogEmpty-4                          380ns ± 3%     370ns ± 3%   -2.84%
(p=0.034 n=6+10)
Info-4                              524ns ±10%     500ns ± 7%     ~
(p=0.099 n=6+10)
ContextFields-4                     839ns ± 5%     746ns ± 4%  -11.06%
(p=0.000 n=6+10)
ContextAppend-4                     924ns ± 4%     873ns ±13%   -5.46%
(p=0.034 n=6+9)
LogFields-4                        2.52µs ± 3%    2.38µs ± 2%   -5.44%
(p=0.000 n=6+9)
LogWithNoFields-4                   872ns ± 7%     801ns ± 6%   -8.17%
(p=0.001 n=6+10)
LogWithComplexFields-4             10.8µs ± 5%    10.4µs ± 2%   -3.78%
(p=0.007 n=5+9)
LogWithComplexFieldsInContext-4    2.78µs ±40%    2.23µs ±19%  -19.95%
(p=0.016 n=6+10)

name                             old alloc/op   new alloc/op   delta
LogEmpty-4                           144B ± 0%      144B ± 0%     ~
(all equal)
Info-4                               352B ± 0%      352B ± 0%     ~
(all equal)
ContextFields-4                      896B ± 0%      896B ± 0%     ~
(all equal)
ContextAppend-4                      496B ± 0%      496B ± 0%     ~
(all equal)
LogFields-4                        1.52kB ± 0%    1.52kB ± 0%     ~
(all equal)
LogWithNoFields-4                    343B ± 2%      337B ± 2%   -1.87%
(p=0.003 n=5+9)
LogWithComplexFields-4             4.84kB ± 2%    4.79kB ± 0%   -1.10%
(p=0.001 n=6+10)
LogWithComplexFieldsInContext-4    2.18kB ± 4%    2.13kB ± 1%   -2.21%
(p=0.019 n=5+8)

name                             old allocs/op  new allocs/op  delta
LogEmpty-4                           3.00 ± 0%      3.00 ± 0%     ~
(all equal)
Info-4                               5.00 ± 0%      5.00 ± 0%     ~
(all equal)
ContextFields-4                      7.00 ± 0%      7.00 ± 0%     ~
(all equal)
ContextAppend-4                      12.0 ± 0%      12.0 ± 0%     ~
(all equal)
LogFields-4                          30.0 ± 0%      30.0 ± 0%     ~
(all equal)
LogWithNoFields-4                    4.00 ± 0%      4.00 ± 0%     ~
(all equal)
LogWithComplexFields-4               59.0 ± 0%      59.0 ± 0%     ~
(all equal)
LogWithComplexFieldsInContext-4      14.0 ± 0%      14.0 ± 0%     ~
(all equal)

tgulacsi committed:
testing.B does not have a Short method

This compile error is hidden by the "lookup" tag.

dglsparsons commented on bench_lookup_test.go:
```diff
@@ -0,0 +1,59 @@
+// +build lookup
+
+package antilog_test
```
This isn't testing anything antilog related. Could it be removed?

dglsparsons commented on fields.go:
```diff
@@ -21,21 +21,38 @@ type EncodedFields []EncodedField
+			length := len(f)
+			offset = len(fields)
+			res = make(EncodedFields, offset+length)
+			copy(res[offset:], f)
```
I think you could reassign res[offset:] to `f` here, avoiding the `res == nil` check below, meaning you can always return `f`

dglsparsons replied:
It should also remove the need to assign `f` or `res` to `flds`

tgulacsi replied:
Godd catch, thanks for the idea!

dglsparsons commented on fields.go:
```diff
@@ -21,21 +21,38 @@ type EncodedFields []EncodedField
 	for ix := len(fields) - 1; ix >= 0; ix-- {
 		field := fields[ix]
-
+		key := field.Key()
```
This looks like an unnecessary assignment

tgulacsi replied:
It can be moved down to line 32, but it is used in the search loop.

dglsparsons replied:
yeah - it's assigned to `key` and only used in the one place i mean, and i don't think it aids the readability over just using `if v.Key() == field.Key()` directly.

tgulacsi replied:
It is given it's own variable to not have to call Key method on every iteration of the search loop.

dglsparsons commented on fields.go:
```diff
@@ -21,21 +21,38 @@ type EncodedFields []EncodedField
 // PrependUnique encoded field if the key is not already set
 func (f EncodedFields) PrependUnique(fields EncodedFields) EncodedFields {
+	var res EncodedFields
+	var offset int
```
You could initialise `offset` here to `len(fields)`

dglsparsons commented on log.go:
```diff
@@ -50,21 +50,29 @@ func (a AntiLog) Write(msg string, fields ...Field) {
 	}
-
-	fmt.Fprintln(a.Writer, encodedFieldsToJSONObject(encodedFields))
+	w.Write(sb.Bytes())
```
You should probably dogsled this, as it has return values. 
```
_, _ = w.Write(...)
```

dglsparsons replied:
(so it's explicit the error is being ignored)

dglsparsons commented on log.go:
```diff
@@ -50,21 +50,29 @@ func (a AntiLog) Write(msg string, fields ...Field) {
+	sb.WriteString(`{ "timestamp": "`)
+	sb.WriteString(now.Format(time.RFC3339))
+	sb.WriteString(`", "message": `)
+	json.NewEncoder(&sb).Encode(msg)
```
As below - should probably explicitly ignore the error here 
```
_ = json.New...
```

dglsparsons commented on benchmark_test.go:
```diff
@@ -0,0 +1,70 @@
+package antilog
```
can we put this into a `antilog_test` package instead

tgulacsi committed:
Remove bench_lookup_test.go

tgulacsi committed:
Simplify EncodedFields.PrependUnique

tgulacsi committed:
Remove extra \n from the end of message

json.Encoder ends with \n unconditionally,
json.Marshal does not.

Also, introduce a buffer pool (sync.Pool) to decrease allocations even
more.

$ go test -bench=. -benchmem
goos: linux
goarch: amd64
pkg: github.com/shamaazi/antilog
BenchmarkLogEmpty-4                              3765540 326 ns/op              32 B/op          1 allocs/op
BenchmarkInfo-4                                  3108966 393 ns/op              48 B/op          2 allocs/op
BenchmarkContextFields-4                         2165385 555 ns/op             176 B/op          3 allocs/op
BenchmarkContextAppend-4                         1494231 750 ns/op             240 B/op          9 allocs/op
BenchmarkLogFields-4                              538449 2135 ns/op             800 B/op         26 allocs/op
BenchmarkLogWithNoFields-4                       1486839 775 ns/op             227 B/op          2 allocs/op
BenchmarkLogWithComplexFields-4                   114956 9915 ns/op            3854 B/op         55 allocs/op
BenchmarkLogWithComplexFieldsInContext-4          580348 2572 ns/op            1927 B/op         10 allocs/op
PASS
ok      github.com/shamaazi/antilog     12.821s

dglsparsons commented:
Thanks for your contribution here - looks like some crazy reduction in allocations. It's much appreciated.
